docs/src/manual/migrate_from_flux.md

```diff
@@ -0,0 +1,213 @@
+# Migrating from Flux to Lux
+
+For the core library layers like [`Dense`](@ref), [`Conv`](@ref), etc. we have intentionlly
+kept the API very similar to Flux. In most cases, replacing `using Flux` with `using Lux`
+should be enough to get you started. We cover the additional changes that you will have to
+make in the following example.
+
+
+```@raw html
+=== "Lux"
+
+    ```julia hl_lines="1 7 9 11"
+    using Lux, Random, NNlib, Zygote
+
+    model = Chain(Dense(2 => 4), BatchNorm(4, relu), Dense(4 => 2))
+    rng = Random.default_rng()
+    x = randn(rng, 2, 4)
+  
+    ps, st = Lux.setup(rng, model)
+
+    model(x, ps, st)
+
+    gradient(ps -> sum(first(model(x, ps, st))), ps)
+    ```
+
+=== "Flux"
+
+    ```julia
+    using Flux, Random, NNlib, Zygote
+
+    model = Chain(Dense(2 => 4), BatchNorm(4, relu), Dense(4 => 2))
+    rng = Random.default_rng()
+    x = randn(rng, 2, 4)
+
+
+
+    model(x)
+
+    gradient(model -> sum(model(x)), model)
+    ```
+```
+
+## Implementing Custom Layers
+
+Flux and Lux operate under extremely different design philosophies regarding how layers
+should be implemented. A summary of the differences would be:
+
+* Flux stores everything in a single struct and relies on `Functors.@functor` and
+  `Flux.trainable` to distinguish between trainable and non-trainable parameters.
+
+* Lux relies on the user to define [`Lux.initialparameters`](@ref) and
+  [`Lux.initialstates`](@ref) to distinguish between trainable parameters (called
+  "parameters") and non-trainable parameters (called "states"). Additionally Lux layers
+  define the model architecture, hence device transfer utilities like [`gpu`](@ref),
+  [`cpu`](@ref), etc. cannot be applied on Lux layers, instead they need to be applied on
+  the parameters and states.
+
+Let's work through a concrete example to demonstrate this. We will implement a very simple
+layer that computes ``A \times B \times x`` where ``A`` is not trainable and ``B`` is
+trainable.
+
+```@raw html
+=== "Lux"
+
+    ```julia
+    using Lux, Random, NNlib, Zygote
+
+    struct LuxLinear <: Lux.AbstractExplicitLayer
+        init_A
+        init_B
+    end
+
+    function LuxLinear(A::AbstractArray, B::AbstractArray)
+        # Storing Arrays or any mutable structure inside a Lux Layer is not recommended
+        # instead we will convert this to a function to perform lazy initialization
+        return LuxLinear(() -> copy(A), () -> copy(B))
+    end
+
+    # `B` is a parameter
+    Lux.initialparameters(rng::AbstractRNG, layer::LuxLinear) = (B=layer.init_B(),)
+
+    # `A` is a state
+    Lux.initialstates(rng::AbstractRNG, layer::LuxLinear) = (A=layer.init_A(),)
+
+    (l::LuxLinear)(x, ps, st) = st.A * ps.B * x, st
+    ```
+
+=== "Flux"
+
+    ```julia
+    using Flux, Random, NNlib, Zygote
+
+    struct FluxLinear
+        A
+        B
+    end
+
+
+
+
+
+
+
+    # `A` is not trainable
+    Flux.trainable(f::FluxLinear) = (f.B,)
```

Comment:
What you wrote is the way to go (default to `@functor [Layer] [fields...]` unless you have non-trainable state, in which case `trainable` is necessary as well (BTW this is exported by Optimisers now, so you can define layers without importing Flux at all. Perhaps that should be the recommendation).

The only change to make is this:
```suggestion
    Flux.trainable(f::FluxLinear) = (; B=f.B)
```
Not including prop names in `trainable` is deprecated and will be removed in the next major version of Optimisers.

Reply:
(1 hour during a work day lunch break is a little short notice for reviewing a PR 😂 )

Reply:
I needed the manual to go through for a corresponding DiffEqSensitivity PR. I will update the example in a new PR.